smithy-aws-apigateway-openapi/src/main/java/software/amazon/smithy/aws/apigateway/openapi/CloudFormationSubstitution.java

```diff
@@ -45,7 +45,7 @@
 
     private static final Logger LOGGER = Logger.getLogger(CloudFormationSubstitution.class.getName());
     private static final String SUBSTITUTION_KEY = "Fn::Sub";
-    private static final Pattern SUBSTITUTION_PATTERN = Pattern.compile("\\$\\{.+}");
+    private static final Pattern SUBSTITUTION_PATTERN = Pattern.compile("\\$\\{[A-Za-z0-9.:]+}");
```

Comment:
I think this is too restrictive. `[\w:]+` would be a little better, though maybe we want hyphens too. Generally I'm just a little nervous about doing what is a potentially breaking change. Could you just specifically exclude that token pattern? For instance, [`\$\{(?!Token\[[^\[\]]*\])[^\{\}]+\}`](regexr.com/60l7a). (You'd need to add `Pattern.CASE_INSENSITIVE` to prevent it from being even more grim.)

Reply:
I've based it on the rules in CFN.
Parameter names: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#parameters-section-structure-requirements - _Each parameter must be given a logical name (also called logical ID), which must be alphanumeric_
Pseudo parameters: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html
Resource logical IDs: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resources-section-structure.html - _The logical ID must be alphanumeric (A-Za-z0-9)_
Resource attributes: Couldn't find it in docs, but I believe they can be only alphanumeric too.

Agree the code change is going from looser to restrictive which is technically backwards incompatible, but given the above, anything else this code allowed earlier, would still have have worked end-to-end.

Reply:
Cool, so other values never would have worked. Thanks for the context!